### ISSUE TITLE
[$600] drawing-tests: improve and refactor

### ISSUE THREAD

**Issue by dkochmanski** [McCLIM/McCLIM#580]
Drawing tests are useful but somewhat hard to navigate. Suggested improvements:

- [x] divide list on the left in two: test category and tests in said category.
For instance: `Text` category when selected has tests in second list (example based on actual list): `Test1`, `Test2`, `Align`, `Clipping` etc.

- [x] design and add macro which simplifies defining drawing method with category / name / description assignment.
Defined method should be possible to provide as a display function in application frame definition and allow accessing the method (i.e `(get-display "Text" "Test1")`).

- [x] option to run each test in a separate application-frame. it should be possible to pick a backend 
Predefiend options include clx-fb and clx backends, but it should be possible to provide a custom server-path. It should be customizable whenever created pane records its output or not. Commands assigned to keys should be added: space = repaint, r = redisplay

- [x] At least two layouts which are easily switachable: backend preview and side-by-side backend+render preview (backend-only should be default)

- [x] Each test should have a separate option to run with different backends (especially ps, pdf and png), right now we have one run for all (which should be preserved as well) - file name should default to something in /tmp but it should be customizable (i.e text field)

- [x] Each test should be elligible to run multiple times for benchmark (see benchmarking demo on how we could measure time).

- [x] Other visual and ux improvements are very welcome.

- [x] tests in a list should be navigable with keyboard arrows

- [x] in print all tests some kind of visual progress bar should be displayed (may be a log information in a pane printed after eac htest)

- [x] in print all tests description should be included as part of the rendered drawing

I plan to assign $600 bounty to this ticket before the upcoming release, but if it is completed earlier, bounty will be added anyway and reward will be paid. We are still working on the requirements though.

**Comment by dkochmanski** [comment]
https://www.bountysource.com/issues/65242857-drawing-tests-improve-and-refactor

**Comment by josrr** [comment]
Hi. I think I have covered several of the requested changes: 1,2,3,4 and 8; the code can be seen [here](https://github.com/josrr/McCLIM/blob/master/Examples/drawing-tests.lisp) and [here](https://www.rufina.link/archivos/drawing-test01.png) is a screenshot. 
I'm working on the other points, but I'd like to know if I'm on the right track.
Also, It seems to be some problem with the list-pane, it only highlight the first item.

**Comment by dkochmanski** [comment]
Hey, thanks for working on that. This looks good, still I have some remarks regarding existing points. I focus on visual appearance for now (I won't look at the code yet):

1. [split categories] It would be better if category were a list too (when item is selected it populates the second list).
2. *requires looking at code*
3. [run on custom port / separate application frame] **This is good.** Little nitpick: text field should have some space after options pane.
4. I'd imagine layout switch as a toggle button (not a menu) `[ ] side-by-side view with the renderer output`
8. **Screenshots doesn't show this one**

Regarding unvisible list elements, please try upgrading your sbcl. We had report like this and it seemed to be issue with this particular implementation (and particular version). If problem persists please report it on the issue tracker.

To sum up: you are on the right track. Don't hestitate to ask when you are in doubt. This is a very good idea to share partial progress.

**Comment by josrr** [comment]
I'm using sbcl 1.4.15; I will report the issue; I have not worked on 5 item yet.

**Comment by josrr** [comment]
New screenshot. 

![drawing-test02](https://user-images.githubusercontent.com/3723736/51066821-639c1680-15d2-11e9-8d30-4a316510e4d2.png)

**Comment by josrr** [comment]
Hello, maybe I rushed doing the pull request; If my code is not acceptable, I would like to know, maybe I can improve it if you tell me what the problem is.

**Comment by dkochmanski** [comment]
Hey, there is no problem on your end. The silence from my part is because I have a little daughter who appeared around that time and I didn't have time yet to give a proper peer review. I'll try to get to it next week if nobody else will go through it. Sorry for keeping you waiting.

**Comment by josrr** [comment]
Hi Daniel, thanks for the reply; congratulations for your new daughter.

**Comment by lokedhs** [comment]
Some issues:

- When clicking on the catory point and the clipping test, the size of the frame changes
- Arrow keys always modifies the "Tests" list. However, it could be argued that this functionality should be a property of the widget itself so it works everywhere. In any case, lists cannot have input focus at the moment so this is the best we can get, I think. Any suggestions how it can be made system-wide is welcome.
- While printing all as PNG, the output isn't updated correctly (see screenshot below)

**Comment by lokedhs** [comment]
Apparently the text drawing error is unrelated to this pull request.

**Comment by josrr** [comment]
I did some tests about the resizing of the frame with the clipping tests; it seems to happen with points, lines, ellipses, rectangles, circles and text; but not with polygons. One way I can solve it is by doing something like this:

```common-lisp
(test-simple-clipping-region stream
                               #'(lambda (stream)
                                   (let* ((thickness (random 100))
                                          (radius (floor thickness 2)))
                                     (clim:draw-point* stream
                                                       (random (- *width* radius))
                                                       (random (- *height* radius))
                                                       :ink (make-random-col)
                                                       :line-thickness thickness))))
```

But, I think I should not have to do that; I think the option `:clipping-region` is not doing a good job.

**Comment by dkochmanski** [comment]
please report it as a separate issue with a minimal (standalone) example to reproduce it

**Comment by dkochmanski** [comment]
I've looked into the drawing tests changes and they look good. I had a few remarks.

**Comment by josrr** [comment]
It seems to me than only points 1,7 and 8 of your remarks are missing to address.
I can not find a way to redraw the backend pane at the end of the layout change. The render pane seems to work because it is recording the pattern where everything is drawn, if it is not recorded it behaves in the same way as the backend pane. ¿Any ideas?

**Comment by dkochmanski** [comment]
this is fixed by a pull request merge. please claim the bounty.
